trieste/space.py

```diff
@@ -329,3 +346,124 @@ def __mul__(self, other: Box) -> Box:
 
     def __deepcopy__(self, memo: dict[int, object]) -> Box:
         return self
+
+
+class ProductSearchSpace(SearchSpace):
+    r"""
+    Product :class:`SearchSpace` consisting of a product of
+    multiple :class:`SearchSpace`s. This class provides functionality for
+    accessing either the resulting combined search space or each individual space.
+
+    Note that this class assumes that individual points in product spaces are
+    represented with their inputs in the same order as specified when initializing
+    the space.
+    """
+
+    def __init__(self, spaces: list[tuple[str, SearchSpace]]):
```

Comment:
This syntax isn't very Pythonic. How about the following?

```suggestion
    def __init__(self, spaces: Sequence[SearchSpace], tags: Optional[Sequence[str]] = None):
```

You would then write TaggedProductSearchSpace([space_A, space_B], ["A", "B"]), which is I think is cleaner. The initialiser can verify that the lists have the same lengths (just as it already verifies tag uniqueness).

You could then also provide default tags (e.g. "0", "1", etc) in the cases where this doesn't matter. Which raises the possibility of making `SearchSpace.__mul__` handle multiplying arbitrary search spaces, not just those of the same kind, falling back to returning a TaggedProductSearchSpace([self, other]) if there's no better impementation.

?

---

If you don't like spitting the tags and spaces, then another alternative is:

```suggestion
    def __init__(self, spaces: OrderedDict[str, SearchSpace]):
```

which I think is a bit cleaner than using lists of tuples (plus it checks tag uniqueness for you).

Reply:
these are all good suggestions!

I will split up spaces and tags, so that we can get the default behaviour. If we go down the orderedDict route, then we cannot do the searchspace.__mul__